bip-musig2/gen_vectors_helper.py

```diff
@@ -17,4 +17,111 @@ def gen_key_agg_vectors():
+            "nonce_indices": [0, 3],
+            "key_indices": [0, 2],
+            "tweak_indices": [0],
+            "is_xonly": [False],
+            "psig_indices": [4, 5],
+        }, {
+            "nonce_indices": [0, 4],
+            "key_indices": [0, 3],
+            "tweak_indices": [0, 1, 2],
+            "is_xonly": [True, False, True],
+            "psig_indices": [6, 7],
+        },
+    ]
+    for (i, test_case) in enumerate(valid_test_cases):
+        is_xonly = test_case["is_xonly"]
+        nonce_indices = test_case["nonce_indices"]
+        key_indices = test_case["key_indices"]
+        psig_indices = test_case["psig_indices"]
+        vec_pnonces = [pnonces[i] for i in nonce_indices]
+        vec_pubkeys = [pubkeys[i] for i in key_indices]
+        vec_tweaks = [tweaks[i] for i in test_case["tweak_indices"]]
+
+        aggnonce = nonce_agg(vec_pnonces)
+        test_case["aggnonce"] = aggnonce.hex().upper()
+        session_ctx = SessionContext(aggnonce, vec_pubkeys, vec_tweaks, is_xonly, msg)
+
+        for j in range(len(key_indices)):
+            psigs[psig_indices[j]] = sign(secnonces[nonce_indices[j]], sk[key_indices[j]], session_ctx)
+        sig = partial_sig_agg([psigs[i] for i in psig_indices], session_ctx)
+        keygen_ctx = key_agg_and_tweak(vec_pubkeys, vec_tweaks, is_xonly)
+        # To maximize coverage of the sig_agg algorithm, we want one public key
+        # point with an even and one with an odd Y coordinate.
+        if i == 0:
+            assert(has_even_y(keygen_ctx[0]))
+        if i == 1:
+            assert(not has_even_y(keygen_ctx[0]))
+        aggpk = get_xonly_pk(keygen_ctx)
+        assert schnorr_verify(msg, aggpk, sig)
+        test_case["expected"] = sig.hex().upper()
+
+    error_test_case = {
+        "nonce_indices": [0, 4],
+        "key_indices": [0, 3],
+        "tweak_indices": [0, 1, 2],
+        "is_xonly": [True, False, True],
+        "psig_indices": [7, 8],
+        "error": {
+            "type": "invalid_contribution",
+            "signer": 1
+        },
+        "comment": "Partial signature is invalid because it exceeds group size"
+    }
+
+    psigs[8] = bytes.fromhex("FFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141")
+
+    vec_pnonces = [pnonces[i] for i in error_test_case["nonce_indices"]]
+    aggnonce = nonce_agg(vec_pnonces)
+    error_test_case["aggnonce"] = aggnonce.hex().upper()
+
+    def tohex_all(l):
+        return list(map(lambda e: e.hex().upper(), l))
+
+    print(" ", json.dumps({ "pubkeys": tohex_all(pubkeys) }))
+    print(" ", json.dumps({ "pnonces": tohex_all(pnonces) }))
+    print(" ", json.dumps({ "tweaks": tohex_all(pnonces) }))
+    print(" ", json.dumps({ "psigs": tohex_all(psigs) }))
+    print(" ", json.dumps({ "msg": msg.hex().upper() }))
+    print(" ", "valid_test_cases:")
+    for test_case in valid_test_cases:
+        print("   ", test_case)
+    print(" ", "error_test_cases:")
+    print("   ", error_test_case)
```

Comment:
Maybe something like this instead? It's also possible to match the key order of the existing JSON by putting a placeholder `"aggnonce": None` at the top of all test cases.

```suggestion
    print(json.dumps({
        "pubkeys": tohex_all(pubkeys),
        "pnonces": tohex_all(pnonces),
        "tweaks": tohex_all(tweaks),
        "psigs": tohex_all(psigs),
        "msg": msg.hex().upper(),
        "valid_test_cases": valid_test_cases,
        "error_test_cases": [error_test_case]
    }, indent=4))
```

Reply:
Oh nice, I had somehow thought that the json dump isn't pretty printed.